staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -853,15 +936,103 @@ func (s *store) ttlOpts(ctx context.Context, ttl int64) ([]clientv3.OpOption, er
 	return []clientv3.OpOption{clientv3.WithLease(id)}, nil
 }
 
+// If key prefix matches the following, the 4th segment will be tenant
+var regexPrefixToCheck = []string{
+	"services/specs/",
+	"services/endpoints/",
+	"apiregistration.k8s.io/apiservices/",
+	"apiextensions.k8s.io/customresourcedefinitions/",
+}
+
 // Based on the key tree structure, figure out which client it needs to go to
 // If there is no data cluster, all goes to system cluster
+// Aside from pathPrefix,
+//
+// 1. If the key has <= 2 segments (split by "/", the first segment is ""), it goes to system cluster
+// 2. If the key following prefix, the 3th segment will be reported as tenant
+//		"services/specs/"
+//		"services/endpoints/"
+//		"apiregistration.k8s.io/apiservices/"
+//		"apiextensions.k8s.io/customresourcedefinitions/"
+// 3. For the rest, the 2th segment will be reported as tenant
+// 4. If tenant name is not found from tenant map, goes to system cluster
 func (s *store) getClientFromKey(key string) *clientv3.Client {
-	if len(s.dataClusterClients) == 0 {
+	// remove prefix
+	lenPrefix := len(s.pathPrefix)
+	if lenPrefix > 0 && s.pathPrefix[lenPrefix-1:lenPrefix] != "/" {
+		lenPrefix++
+	}
+
+	concisedKey := key[lenPrefix:]
+
+	segs := strings.Split(concisedKey, "/")
+	message := fmt.Sprintf("key [%s] segments %v len %d", key, segs, len(segs))
+
+	if len(segs) <= 2 {
+		klog.V(3).Infof("system client: key segments len <= 2. %s ", message)
+		return s.client
+	}
+
+	tenant := ""
+	hasPrefix := false
+	for _, prefix := range regexPrefixToCheck {
+		isMatched, err := regexp.MatchString(prefix, key)
+		if err == nil && isMatched {
+			hasPrefix = true
+			tenant = getTenantForKey(segs, 2)
+		}
+		if err != nil {
+			klog.Errorf("Regex match error %v. key %s", err, key)
+		}
+	}
+	if !hasPrefix {
+		tenant = getTenantForKey(segs, 1)
+	}
+	if tenant == "" {
```

Comment:
We don't think we have empty tenant now. Do we need to add warnf here?

Reply:
We don't have empty tenant. But the algorithm is not intelligent enough to find tenant only. It is possible that there are an extra / in the key so that tenant name was extracted wrong. "" is safe guard.